src/csharp/Grpc.Core/Internal/NativeExtension.cs

```diff
@@ -117,7 +114,43 @@ private static NativeMethods LoadNativeMethods()
             {
                 return LoadNativeMethodsXamarin();
             }
-            return new NativeMethods(LoadUnmanagedLibrary());
+            if (PlatformApis.IsNetCore)
+            {
+                // On .NET Core, native libraries are a supported feature and the SDK makes
+                // sure that the native library is made available in the right location and that
+                // they will be discoverable by the [DllImport] default loading mechanism,
+                // even in some of the more exotic situations such as single file apps.
+                //
+                // While in theory, we could just [DllImport("grpc_csharp_ext")] for all the platforms
+                // and operating systems, the native libraries in the nuget package
+                // need to be laid out in a way that still allows things to work well under
+                // the legacy .NET Framework (where native libraries are a concept unknown to the runtime).
+                // Therefore, we use several flavors of the DllImport attribute
+                // (e.g. the ".x86" vs ".x64" suffix) and we choose the one we want at runtime.
+                // The classes with the list of DllImport'd methods are code generated,
+                // so having more than just one doesn't really bother us.
+
+                // on Windows, the DllImport("grpc_csharp_ext.x64") doesn't work for some reason,
```

Comment:
This was bug https://github.com/dotnet/coreclr/pull/17505 . It is fixed in .NET Core 3.1, but I assume you still want this to work on .NET Core 2.1.